### DISCUSSION TITLE
$dispatch not working in function component

### DISCUSSION THREAD

**Discussion by iivvaannxx** [alpinejs/alpine#3543] [7. Bugs (archived)] [answered]
I am trying to dispatch a custom event from a function component registered through the `Alpine.data()` callback, but it doesn't work and I don't know if I am missing something or what is happening but at this point I can't see anything else than a bug, so it's why I am posting it as a bug.

I prepared a very minimal example that looks like so:
File `foo.js`:

```js
export default function Foo () {

  return {

    init () {

      console.log('foo')
      this.$dispatch('bar')
    },

    text: 'Hello World!',

    events: {

      '@bar' () {

        console.log('bar')
      }
    }
  }
}

```

Somewhere during my website initialisation:
```js
import Foo from 'foo.js'
Alpine.data('foo', Foo)
```

And finally on the HTML markup:
```html
<span x-data = 'foo' x-text='text' x-bind='events' @click = '$dispatch("bar")'></span>
```

If I click the element, the `bar` event gets dispatched correctly and 'bar' gets printed to the console. However, the dispatch done through the `init` callback does not work. The `this` object is correct, because I can access all the other elements and magics like the `$el` or the `$refs`, but the `$dispatch` of the custom event does not work. 

In this example I try the `$dispatch` from the `init` hook but no matter where or when I call it, there's no success at all. I am claryfing this because I don't know if the `init` executes somehow before the `$dispatch` can work.

What I am missing? Is this expected behaviour?

**Comment by SimoTod** [comment, marked as answer]
You are dispatching the event too early. X-init runs before the click listener gets created. If the listener is on the same elemeny, use $nextTick in your x-init

**Reply by iivvaannxx** [reply]
Don't know why, but now it works. I tried waiting before dispatching the event previously and didn't work, guess because I had some other error. My fault I did not try it again. It always happens when I ask for help that it's the most trivial solution ever and I end up feeling dumb.

Anyway, thanks for your help! 😄
